flink-connectors/flink-connector-pubsub/src/main/java/org/apache/flink/streaming/connectors/pubsub/PubSubSink.java

```diff
@@ -0,0 +1,270 @@
+import org.apache.flink.streaming.api.functions.sink.RichSinkFunction;
+import org.apache.flink.streaming.api.functions.sink.SinkFunction;
+import org.apache.flink.streaming.connectors.pubsub.common.SerializableCredentialsProvider;
+
+import com.google.api.gax.core.CredentialsProvider;
+import com.google.api.gax.grpc.GrpcTransportChannel;
+import com.google.api.gax.rpc.FixedTransportChannelProvider;
+import com.google.api.gax.rpc.TransportChannel;
+import com.google.auth.Credentials;
+import com.google.cloud.pubsub.v1.Publisher;
+import com.google.protobuf.ByteString;
+import com.google.pubsub.v1.ProjectTopicName;
+import com.google.pubsub.v1.PubsubMessage;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A sink function that outputs to PubSub.
+ *
+ * @param <IN> type of PubSubSink messages to write
+ */
+public class PubSubSink<IN> extends RichSinkFunction<IN> {
+	private static final Logger LOG = LoggerFactory.getLogger(PubSubSink.class);
+
+	private final SerializableCredentialsProvider serializableCredentialsProvider;
+	private final SerializationSchema<IN> serializationSchema;
+	private final String projectName;
+	private final String topicName;
+	private final String hostAndPort;
+
+	private transient Publisher publisher;
+
+	private PubSubSink(
+		SerializableCredentialsProvider serializableCredentialsProvider,
+		SerializationSchema<IN> serializationSchema,
+		String projectName,
+		String topicName,
+		String hostAndPort) {
+		this.serializableCredentialsProvider = serializableCredentialsProvider;
+		this.serializationSchema = serializationSchema;
+		this.projectName = projectName;
+		this.topicName = topicName;
+		this.hostAndPort = hostAndPort;
+	}
+
+	private transient ManagedChannel managedChannel = null;
+	private transient TransportChannel channel = null;
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		Publisher.Builder builder = Publisher
+			.newBuilder(ProjectTopicName.of(projectName, topicName))
+			.setCredentialsProvider(serializableCredentialsProvider);
+
+		if (hostAndPort != null) {
+			managedChannel = ManagedChannelBuilder
+				.forTarget(hostAndPort)
+				.usePlaintext(true) // This is 'Ok' because this is ONLY used for testing.
+				.build();
+			channel = GrpcTransportChannel.newBuilder().setManagedChannel(managedChannel).build();
+			builder.setChannelProvider(FixedTransportChannelProvider.create(channel));
+		}
+
+		publisher = builder.build();
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+		shutdownPublisher();
+		shutdownTransportChannel();
+		shutdownManagedChannel();
+	}
+
+	private void shutdownPublisher() {
+		try {
+			publisher.shutdown();
+		} catch (Exception e) {
+			LOG.info("Shutting down Publisher failed: {}", e.getMessage());
```

Comment:
Can you do the following for exception logging everywhere in this PR?
```java
LOG.info("Shutting down Publisher failed", e);
```

When running the code, I'm getting log statements like this
```
17:46:49,970 INFO  org.apache.flink.streaming.connectors.pubsub.PubSubSink       - Shutting down Publisher failed: null
17:46:49,970 INFO  org.apache.flink.streaming.connectors.pubsub.PubSubSink       - Shutting down Publisher failed: null
17:46:49,970 INFO  org.apache.flink.streaming.connectors.pubsub.PubSubSink       - Shutting down Publisher failed: null
17:46:49,970 INFO  org.apache.flink.streaming.connectors.pubsub.PubSubSink       - Shutting down Publisher failed: null
17:46:49,970 INFO  org.apache.flink.streaming.connectors.pubsub.PubSubSink       - Shutting down Publisher failed: null
17:46:49,970 INFO  org.apache.flink.streaming.connectors.pubsub.PubSubSink       - Shutting down Publisher failed: null
17:46:49,970 INFO  org.apache.flink.streaming.connectors.pubsub.PubSubSink       - Shutting down Publisher failed: null
17:46:49,970 INFO  org.apache.flink.streaming.connectors.pubsub.PubSubSink       - Shutting down Publisher failed: null
```

With proper exception forwarding, I'd be able to see what's going on.

Reply:
Fixed